smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/TraceFile.java

```diff
@@ -0,0 +1,390 @@
+    public static final String DEFINITIONS_TEXT = "definitions";
+    public static final String SHAPES_TEXT = "shapes";
+    public static final String SMITHY_TRACE_VERSION = "1.0";
+
+    private String smithyTrace;
+    private ArtifactMetadata artifactMetadata;
+    private ArtifactDefinitions artifactDefinitions; //Optional
+    private Map<ShapeId, List<ShapeLink>> shapes;
+    private SourceLocation sl = new SourceLocation("");
+
+    private TraceFile(Builder builder) {
+        smithyTrace = SmithyBuilder.requiredState(SMITHY_TRACE_TEXT, builder.smithyTrace);
+        artifactMetadata = SmithyBuilder.requiredState(ARTIFACT_TEXT, builder.artifactMetadata);
+        shapes = SmithyBuilder.requiredState(SHAPES_TEXT, MapUtils.copyOf(builder.shapes));
+        artifactDefinitions = builder.artifactDefinitions;
+    }
+
+    /**
+     * Parses and validates the trace file passed in as filename
+     * and instantiates smithyTrace and defintions, and fills
+     * artifactMetadata and shapes.
+     *
+     * @param filename the absolute or relative path of tracefile
+     * @throws FileNotFoundException if filename is not found for reading
+     */
+    public static TraceFile parseTraceFile(URI filename) throws FileNotFoundException {
+        InputStream stream = new FileInputStream(new File(filename));
+        ObjectNode node = (ObjectNode) Node.parse(stream);
+        return createFromNode(node);
+    }
+
+    /**
+     * Converts ObjectNode into TraceFile.
+     *
+     * @param value an ObjectNode that represents the entire trace file.
+     */
+    public static TraceFile createFromNode(Node value) {
+        ObjectNode node = value.expectObjectNode();
+        Builder builder = builder()
+                .smithyTrace(node.expectStringMember(SMITHY_TRACE_TEXT).getValue())
+                .artifact(ArtifactMetadata.createFromNode(node.expectObjectMember(ARTIFACT_TEXT)));
+
+        //parse shapes
+        Map<StringNode, Node> shapeMap = node.expectObjectMember(SHAPES_TEXT).getMembers();
+        for (Map.Entry<StringNode, Node> entry : shapeMap.entrySet()) {
+            for (Node linkNode : (entry.getValue().expectArrayNode()).getElements()) {
+                builder.addShapeLink(entry.getKey().getValue(), ShapeLink.createFromNode(linkNode));
+            }
+        }
+
+        //parse definitions
+        if (node.containsMember(DEFINITIONS_TEXT)) {
+            builder.definitions(ArtifactDefinitions.createFromNode(node.expectObjectMember(DEFINITIONS_TEXT)));
+        }
+
+        return builder.build();
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    /**
+     * Writes the TraceFile JSON ObjectNode
+     * to construct a trace file.
+     *
+     * @param fileName absolute or relative path to write the trace file
+     * @throws IOException if there is an error writing to fileName
+     */
+    public void writeTraceFile(String fileName) throws IOException {
```

Comment:
Probably should remove this, see the previous comment about separation of concerns with file loading.